### ISSUE TITLE
EF Core 5.x and later, cannot use Change-tracking proxies with EntityFrameworkDataProvider

### ISSUE THREAD

**Issue by CyberBotX** [thepirat000/Audit.NET#468]
**Describe the bug**
Since EF Core 5.0, Microsoft has put built-in support for change-tracking proxies via the Microsoft.EntityFrameworkCore.Proxies package, as described here: https://docs.microsoft.com/en-us/ef/core/change-tracking/change-detection#change-tracking-proxies

When trying to use change-tracking proxies with the EntityFrameworkDataProvider, I get an exception on trying to save changes to the database because the object instance of the audit entity was created in Audit.NET using Activator.CreateInstance instead of using the DbContext's CreateProxy. Specifically the exception is as follows:

`System.InvalidOperationException: 'The entity type 'AuditLog' is configured to use the 'ChangingAndChangedNotifications' change tracking strategy, but does not implement the required 'INotifyPropertyChanging' interface. Implement 'INotifyPropertyChanging' on 'AuditLog' or use a different change tracking strategy.'
`

**To Reproduce**
(This is just a pretty bare bones set of classes, which I hope should be enough to demonstrate the problem.)
```
public class AuditLog
{
    public int Id { get; set; }
    public DateTime DateTime { get; set; }
}
public class Test
{
    public int Id { get; set; }
    public string TestString { get; set; }
}
public class MyDbContext : AuditDbContext
{
    public DbSet<Test> Tests => this.Set<Test>();

    // Assume there is an OnConfiguring here that calls UseChangeTrackingProxies() on the optionsBuilder
}
class Program
{
    static void Main()
    {
        _ = Audit.Core.Configuration.Setup().UseEntityFramework(ef => ef.
            UseDbContext<MyDbContext>().
            AuditTypeMapper(t => typeof(AuditLog)).
            AuditEntityAction<AuditLog>((ev, ent, auditEntity) => auditEntity.DateTime = DateTime.Now).
            IgnoreMatchedProperties(true)
        );
        using var context = new MyDbContext();
        _ = context.Test.Add(context.CreateProxy<Test>(test => test.TestString = "Test"));
        _ = context.SaveChanges(); // <-- Exception happens when this is called
    }
}
```

**Expected behavior**
Audit.NET's EntityFrameworkDataProvider should detect that the DbContext uses change-tracking proxies and change how it creates the audit entity for the database.

**Libraries (specify the Audit.NET extensions being used including version):**
- Audit.EntityFramework.Core: 19.0.0
- Microsoft.EntityFrameworkCore.Proxies: 5.0.12
(I suspect that the type of database being used won't matter)

**Target .NET framework:**
- .NET 5.0 (but I suspect the framework won't matter, what matters is if EF Core 5.x or later is being used)

**Additional context**
When not using change-tracking proxies, everything works as expected.

**Comment by thepirat000** [comment]
I think the library _could_ detect if the entity should be proxied with the following comparison `EntityEntry.Metadata.GetChangeTrackingStrategy() != ChangeTrackingStrategy.Snapshot`, and create the entity with `dbContext.CreateProxy(type)`. 

But a new package reference must be included in order to be able to create a proxied entity from the library: `Microsoft.EntityFrameworkCore.Proxies`.

That reference will be there even if you don't use change-tracking proxies...

So, maybe a better option would be to provide a mechanism for the Audit Type creation on the client, something like

`.AuditFactoryMapper(ctx => ctx.CreateProxy<AuditLog>()).`

But still not sure what should be the best option.

**Comment by CyberBotX** [comment]
Personally, that suggestion to add a factory mapper sounds like a pretty good one. Adding the package reference to Audit.NET sounds like overkill if someone isn't using change-tracking proxies.

**Comment by CyberBotX** [comment]
After I got to work, I thought about this a little, perhaps you could have 2 signatures for that factory mapper, one as you suggested which would work for the single-table audit but another that provides both the context and the audit entity type (from the audit type mapper). CreateProxy has an overload which is non-generic and takes the `Type` as an argument.

**Comment by thepirat000** [comment]
Yes, that was just an example, but I think it will have two parameters, the _audit db context_ and the _entity entry_ (from which you can get the type, the table, the original db context, etc...)

**Comment by thepirat000** [comment]
You can now use the `AuditEntityCreator` configuration on the EF data provider in order to specify the way to create the audit entity, instead of mapping it to a type.

There are two overloads, one that takes the AuditDbContext as a parameter, and another one that takes the AuditDbContext and the EntityEntry,

You should be able to configure as follows:

```c#
Audit.Core.Configuration.Setup()
	.UseEntityFramework(ef => ef
		.UseDbContext<YourContext>()
		.AuditEntityCreator(ctx => ctx.CreateProxy<AuditLog>())
```

This was implemented starting on version 19.0.1, please upgrade your references and re-test

**Comment by CyberBotX** [comment]
I can confirm when using `AuditEntityCreator` that things work perfectly as expected, thank you for making this update. :D
